File: backend/app/domains/quality_gates/schemas/quality_gates.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


class QualityGateThresholds(BaseModel):
    """Configurable threshold values for a quality gate.

    All metric fields are optional; only configured ones are enforced.
    P0 thresholds block the deploy unconditionally; P1 thresholds are warnings
    unless the gate's p0_only flag is false (default: all configured thresholds
    are P0).
    """

    retrieval_hit_rate_min: float | None = Field(default=None, ge=0.0, le=1.0)
    citation_accuracy_score_min: float | None = Field(default=None, ge=0.0, le=1.0)
    faithfulness_score_min: float | None = Field(default=None, ge=0.0, le=1.0)
    answer_relevance_score_min: float | None = Field(default=None, ge=0.0, le=1.0)
    not_found_rate_max: float | None = Field(default=None, ge=0.0, le=1.0)
    safety_pass_rate_min: float | None = Field(default=None, ge=0.0, le=1.0)
    latency_ms_p95_max: float | None = Field(default=None, ge=0.0)
    cost_usd_per_question_max: float | None = Field(default=None, ge=0.0)


class CreateQualityGateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=8000)
    thresholds: QualityGateThresholds = Field(default_factory=QualityGateThresholds)
    baseline_evaluation_run_id: str | None = None
    baseline_safety_run_id: str | None = None

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("name must not be blank")
        return trimmed

    @field_validator("description")
    @classmethod
    def validate_description(cls, value: str | None) -> str | None:
        if value is None:
            return value
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("description must not be blank")
        return trimmed


class UpdateQualityGateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None)
    thresholds: QualityGateThresholds | None = None
    baseline_evaluation_run_id: str | None = Field(default=None)
    baseline_safety_run_id: str | None = Field(default=None)

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str | None) -> str | None:
        if value is None:
            return value
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("name must not be blank")
        return trimmed


class QualityGateResponse(BaseModel):
    quality_gate_id: str
    name: str
    description: str | None = None
    thresholds: dict[str, object]
    baseline_evaluation_run_id: str | None = None
    baseline_safety_run_id: str | None = None
    created_by_id: str | None = None
    created_at: datetime
    updated_at: datetime


class QualityGateListResponse(BaseModel):
    items: list[QualityGateResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# Gate run
# ---------------------------------------------------------------------------


class TriggerQualityGateRunRequest(BaseModel):
    evaluation_run_id: str | None = None
    safety_eval_run_id: str | None = None

    @field_validator("evaluation_run_id")
    @classmethod
    def validate_evaluation_run_id(cls, value: str | None) -> str | None:
        if value is None:
            return value
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("evaluation_run_id must not be blank")
        return trimmed

    @field_validator("safety_eval_run_id")
    @classmethod
    def validate_safety_eval_run_id(cls, value: str | None) -> str | None:
        if value is None:
            return value
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("safety_eval_run_id must not be blank")
        return trimmed


class GateCheckResult(BaseModel):
    metric: str
    label: str
    threshold: float
    actual: float | None
    passed: bool
    detail: str | None = None


class QualityGateRunResponse(BaseModel):
    gate_run_id: str
    quality_gate_id: str
    evaluation_run_id: str | None = None
    safety_eval_run_id: str | None = None
    verdict: Literal["passed", "failed", "overridden"]
    passed_checks: list[GateCheckResult]
    failed_checks: list[GateCheckResult]
    override_reason: str | None = None
    overridden_by_id: str | None = None
    overridden_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class QualityGateRunListResponse(BaseModel):
    items: list[QualityGateRunResponse]
    total: int
    limit: int
    offset: int


class QualityGateOverrideRequest(BaseModel):
    reason: str = Field(min_length=10, max_length=2000)

    @field_validator("reason")
    @classmethod
    def validate_reason(cls, value: str) -> str:
        trimmed = value.strip()
        if len(trimmed) < 10:
            raise ValueError("reason must be at least 10 characters")
        return trimmed


class QualityGateReportResponse(BaseModel):
    """Full gate report — returned as CI artifact JSON."""

    gate_run_id: str
    quality_gate_id: str
    quality_gate_name: str
    verdict: str
    generated_at: str
    evaluation_run_id: str | None = None
    safety_eval_run_id: str | None = None
    thresholds_applied: dict[str, object]
    passed_checks: list[GateCheckResult]
    failed_checks: list[GateCheckResult]
    total_checks: int
    pass_count: int
    fail_count: int
    override_reason: str | None = None
    overridden_by_id: str | None = None
    overridden_at: str | None = None
    evaluation_summary: dict[str, object] | None = None
    safety_summary: dict[str, object] | None = None
    ci_exit_code: int

File: frontend/src/lib/api/quality-gates.ts
import { apiRequest } from "@/lib/api/request";

export type QualityGateVerdict = "passed" | "failed" | "overridden";

export type QualityGateThresholds = {
  retrieval_hit_rate_min?: number | null;
  citation_accuracy_score_min?: number | null;
  faithfulness_score_min?: number | null;
  answer_relevance_score_min?: number | null;
  not_found_rate_max?: number | null;
  safety_pass_rate_min?: number | null;
  latency_ms_p95_max?: number | null;
  cost_usd_per_question_max?: number | null;
};

export type QualityGateResponse = {
  quality_gate_id: string;
  name: string;
  description: string | null;
  thresholds: Record<string, number | null>;
  baseline_evaluation_run_id: string | null;
  baseline_safety_run_id: string | null;
  created_by_id: string | null;
  created_at: string;
  updated_at: string;
};

export type QualityGateListResponse = {
  items: QualityGateResponse[];
  total: number;
  limit: number;
  offset: number;
};

export type GateCheckResult = {
  metric: string;
  label: string;
  threshold: number;
  actual: number | null;
  passed: boolean;
  detail: string | null;
};

export type QualityGateRunResponse = {
  gate_run_id: string;
  quality_gate_id: string;
  evaluation_run_id: string | null;
  safety_eval_run_id: string | null;
  verdict: QualityGateVerdict;
  passed_checks: GateCheckResult[];
  failed_checks: GateCheckResult[];
  override_reason: string | null;
  overridden_by_id: string | null;
  overridden_at: string | null;
  created_at: string;
  updated_at: string;
};

export type QualityGateRunListResponse = {
  items: QualityGateRunResponse[];
  total: number;
  limit: number;
  offset: number;
};

export type QualityGateReportResponse = {
  gate_run_id: string;
  quality_gate_id: string;
  quality_gate_name: string;
  verdict: string;
  generated_at: string;
  evaluation_run_id: string | null;
  safety_eval_run_id: string | null;
  thresholds_applied: Record<string, number | null>;
  passed_checks: GateCheckResult[];
  failed_checks: GateCheckResult[];
  total_checks: number;
  pass_count: number;
  fail_count: number;
  override_reason: string | null;
  overridden_by_id: string | null;
  overridden_at: string | null;
  evaluation_summary: Record<string, unknown> | null;
  safety_summary: Record<string, unknown> | null;
  ci_exit_code: number;
};

export type CreateQualityGateRequest = {
  name: string;
  description?: string | null;
  thresholds?: QualityGateThresholds;
  baseline_evaluation_run_id?: string | null;
  baseline_safety_run_id?: string | null;
};

export type UpdateQualityGateRequest = {
  name?: string | null;
  description?: string | null;
  thresholds?: QualityGateThresholds | null;
  baseline_evaluation_run_id?: string | null;
  baseline_safety_run_id?: string | null;
};

export type TriggerQualityGateRunRequest = {
  evaluation_run_id?: string | null;
  safety_eval_run_id?: string | null;
};

export type QualityGateOverrideRequest = {
  reason: string;
};

export async function listQualityGates(
  params: { limit?: number; offset?: number } = {},
): Promise<QualityGateListResponse> {
  return apiRequest<QualityGateListResponse>("/quality-gates", {
    query: { limit: params.limit, offset: params.offset },
  });
}

export async function createQualityGate(
  payload: CreateQualityGateRequest,
): Promise<QualityGateResponse> {
  return apiRequest<QualityGateResponse>("/quality-gates", {
    method: "POST",
    json: payload,
  });
}

export async function getQualityGate(gateId: string): Promise<QualityGateResponse> {
  return apiRequest<QualityGateResponse>(
    `/quality-gates/${encodeURIComponent(gateId)}`,
  );
}

export async function updateQualityGate(
  gateId: string,
  payload: UpdateQualityGateRequest,
): Promise<QualityGateResponse> {
  return apiRequest<QualityGateResponse>(
    `/quality-gates/${encodeURIComponent(gateId)}`,
    { method: "PATCH", json: payload },
  );
}

export async function deleteQualityGate(gateId: string): Promise<void> {
  return apiRequest<void>(`/quality-gates/${encodeURIComponent(gateId)}`, {
    method: "DELETE",
  });
}

export async function triggerQualityGateRun(
  gateId: string,
  payload: TriggerQualityGateRunRequest,
): Promise<QualityGateRunResponse> {
  return apiRequest<QualityGateRunResponse>(
    `/quality-gates/${encodeURIComponent(gateId)}/runs`,
    { method: "POST", json: payload },
  );
}

export async function listQualityGateRuns(
  gateId: string,
  params: { limit?: number; offset?: number } = {},
): Promise<QualityGateRunListResponse> {
  return apiRequest<QualityGateRunListResponse>(
    `/quality-gates/${encodeURIComponent(gateId)}/runs`,
    { query: { limit: params.limit, offset: params.offset } },
  );
}

export async function getQualityGateRun(
  gateRunId: string,
): Promise<QualityGateRunResponse> {
  return apiRequest<QualityGateRunResponse>(
    `/quality-gates/runs/${encodeURIComponent(gateRunId)}`,
  );
}

export async function getQualityGateReport(
  gateRunId: string,
): Promise<QualityGateReportResponse> {
  return apiRequest<QualityGateReportResponse>(
    `/quality-gates/runs/${encodeURIComponent(gateRunId)}/report`,
  );
}

export async function overrideQualityGateRun(
  gateRunId: string,
  payload: QualityGateOverrideRequest,
): Promise<QualityGateRunResponse> {
  return apiRequest<QualityGateRunResponse>(
    `/quality-gates/runs/${encodeURIComponent(gateRunId)}/override`,
    { method: "POST", json: payload },
  );
}
